clippy_lints/src/implicit_saturating_sub.rs

```diff
@@ -0,0 +1,90 @@
+use crate::utils::{higher, in_macro, span_lint_and_sugg};
+use if_chain::if_chain;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{BinOpKind, Expr, ExprKind, QPath, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for implicit saturating subtraction.
+    ///
+    /// **Why is this bad?** Simplicity and readability. Instead we can easily use an inbuilt function.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let end: u32 = 10;
+    /// let start: u32 = 5;
+    ///
+    /// let mut i: u32 = end - start;
+    ///
+    /// // Bad
+    /// if i != 0 {
+    ///     i -= 1;
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let end: u32 = 10;
+    /// let start: u32 = 5;
+    ///
+    /// let mut i: u32 = end - start;
+    ///
+    /// // Good
+    /// i = i.saturating_sub(1);
+    /// ```
+    pub IMPLICIT_SATURATING_SUB,
+    pedantic,
+    "Perform saturating subtraction instead of implicitly checking lower bound of data type"
+}
+
+declare_lint_pass!(ImplicitSaturatingSub => [IMPLICIT_SATURATING_SUB]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ImplicitSaturatingSub {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr<'tcx>) {
+        if in_macro(expr.span) {
+            return;
+        }
+        if_chain! {
+            if let Some((ref cond, ref then, None)) = higher::if_block(&expr);
+            // Check if the conditional expression is a binary operation
+            if let ExprKind::Binary(ref op, ref left, ref right) = cond.kind;
+            // Ensure that the binary operator is > or !=
+            if BinOpKind::Ne == op.node || BinOpKind::Gt == op.node;
+            if let ExprKind::Path(ref cond_path) = left.kind;
```

Comment:
The expression compared doesn't have to be a path. In fact, it doesn't matter what the expression is. The only thing that matters is that `expr_ty(expr).is_integral() == true` and that `utils::SpanlessEq` of this expr is the same as of the expr in the if body.